packages/web3-eth/src/utils/get_revert_reason.ts

```diff
@@ -0,0 +1,84 @@
+
+You should have received a copy of the GNU Lesser General Public License
+along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { Web3Context } from 'web3-core';
+import { ContractExecutionError, Eip838ExecutionError, InvalidResponseError } from 'web3-errors';
+import { decodeContractErrorData, isAbiErrorFragment } from 'web3-eth-abi';
+import { AbiErrorFragment, ContractAbi, EthExecutionAPI, TransactionCall } from 'web3-types';
+import { DataFormat, DEFAULT_RETURN_FORMAT } from 'web3-utils';
+
+// eslint-disable-next-line import/no-cycle
+import { call } from '../rpc_method_wrappers';
+import { RevertReason, RevertReasonWithCustomError } from '../types';
+
+/**
+ *	Returns the revert reason generated by the EVM if the transaction were to be executed.
+ *
+ * @param web3Context - ({@link Web3Context}) Web3 configuration object that contains things such as the provider, request manager, wallet, etc.
+ * @param transaction - A transaction object where all properties are optional except `to`, however it's recommended to include the `from` property or it may default to `0x0000000000000000000000000000000000000000` depending on your node or provider.
+ * @returns `undefined` if no revert reason was given, a revert reason object, a revert reason string, or an `unknown` error
+ */
+export async function getRevertReason<
+	ReturnFormat extends DataFormat = typeof DEFAULT_RETURN_FORMAT,
+>(
+	web3Context: Web3Context<EthExecutionAPI>,
+	transaction: TransactionCall,
+	contractAbi?: ContractAbi,
+	returnFormat: ReturnFormat = DEFAULT_RETURN_FORMAT as ReturnFormat,
+): Promise<undefined | RevertReason | RevertReasonWithCustomError | string | unknown> {
+	try {
+		await call(web3Context, transaction, web3Context.defaultBlock, returnFormat);
+		return undefined;
+	} catch (error) {
+		if (
+			error instanceof ContractExecutionError &&
+			error.innerError instanceof Eip838ExecutionError
+		) {
+			if (contractAbi !== undefined) {
+				const errorsAbi = contractAbi.filter(abi =>
+					isAbiErrorFragment(abi),
+				) as unknown as AbiErrorFragment[];
+				decodeContractErrorData(errorsAbi, error.innerError);
+
+				return {
+					reason: error.innerError.message,
+					signature: error.innerError.data?.slice(0, 10),
+					data: error.innerError.data?.substring(10),
+					customErrorName: error.innerError.errorName,
+					customErrorDecodedSignature: error.innerError.errorSignature,
+					customErrorArguments: error.innerError.errorArgs,
+				} as RevertReasonWithCustomError;
+			}
+
+			return {
+				reason: error.innerError.message,
+				signature: error.innerError.data?.slice(0, 10),
+				data: error.innerError.data?.substring(10),
+			} as RevertReason;
+		}
+
+		if (
+			error instanceof InvalidResponseError &&
+			!Array.isArray(error.innerError) &&
+			error.innerError !== undefined
+		) {
+			return error.innerError.message;
+		}
+
+		return error instanceof Error ? error.message : error;
```

Comment:
In case of generic error ( that is not due to failed tx reverted ) , we should throw it and remove `string` from return type`Promise<undefined | RevertReason | RevertReasonWithCustomError | string | unknown>`. 

Because for `getRevertReason` function call with eth_call rpc in case of generic errors: 
1. neither tx eth_call was reverted with reason.
2. nor it succeeded with out any revert reason, 
but request for looking any of above answers failed, that user should catch. It will also make return types homogeneous.

Reply:
Updated via [this commit](https://github.com/web3/web3.js/pull/5844/commits/28a3d848830a0e17aa706b5092c1e917e1b87f02), however `string` is still a possible return type because of

```typescript
if (
	error instanceof InvalidResponseError &&
	!Array.isArray(error.innerError) &&
	error.innerError !== undefined
) {
	return error.innerError.message;
}
```

Reply:
if InvalidResponseError is also not categorizable under 1 and 2, shouldn't it also be thrown?

Reply:
@jdevcs The code snippet above is for errors such as

```typescript
InvalidResponseError: Returned error: err: intrinsic gas too low: have 0, want 21544 (supplied gas 0)
          at Web3RequestManager._processJsonRpcResponse (/home/anon/Public/code/ChainSafe/git-repos/web3.js/packages/web3-core/src/web3_request_manager.ts:372:11)
          at Web3RequestManager.<anonymous> (/home/anon/Public/code/ChainSafe/git-repos/web3.js/packages/web3-core/src/web3_request_manager.ts:216:16)
          at Generator.next (<anonymous>)
          at fulfilled (/home/anon/Public/code/ChainSafe/git-repos/web3.js/packages/web3-core/lib/web3_request_manager.js:5:58)
          at processTicksAndRejections (node:internal/process/task_queues:95:5) {
        innerError: {
          code: -32000,
          message: 'err: intrinsic gas too low: have 0, want 21544 (supplied gas 0)'
        },
        code: 101,
        data: undefined
      }
```

Reply:
@jdevcs That's a good point. I think the error in the chunk parser should remain `InvalidResponseError`, but errors like the gas too low from above should be renamed to something like `RpcError`

If you agree, I can create an issue for it, merge this PR, then make the error name change?

Reply:
sure. Good work @spacesailor24